x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.test.ts

```diff
@@ -0,0 +1,1006 @@
+        expect(CredentialsLogic.values.isCredentialsDataComplete).toEqual(true);
+      });
+    });
+
+    it('should not change any other values', () => {
+      mount();
+      CredentialsLogic.actions.setCredentialsData(meta, [{}, {}]);
+      expect(CredentialsLogic.values).toEqual({
+        ...DEFAULT_VALUES,
+        apiTokens: expect.any(Array),
+        meta: expect.any(Object),
+        isCredentialsDataComplete: expect.any(Boolean),
+      });
+    });
+  });
+
+  describe('setCredentialsDetails', () => {
+    const credentialsDetails = {
+      apiUrl: 'http://example.com',
+      engines: [{}, {}],
+    };
+
+    describe('isCredentialsDataComplete', () => {
+      it('should be set to true so that we know data fetching has been completed', () => {
+        mount({
+          isCredentialsDetailsComplete: false,
+        });
+        CredentialsLogic.actions.setCredentialsDetails(credentialsDetails);
+        expect(CredentialsLogic.values.isCredentialsDetailsComplete).toEqual(true);
+      });
+    });
+
+    describe('engines', () => {
+      it('should set `engines` from the provided details object', () => {
+        mount({
+          engines: [],
+        });
+        CredentialsLogic.actions.setCredentialsDetails(credentialsDetails);
+        expect(CredentialsLogic.values.engines).toEqual([{}, {}]);
+      });
+    });
+
+    describe('apiUrl', () => {
+      it('should set `engines` from the provided details object', () => {
+        mount({
+          apiUrl: '',
+        });
+        CredentialsLogic.actions.setCredentialsDetails(credentialsDetails);
+        expect(CredentialsLogic.values.apiUrl).toEqual('http://example.com');
+      });
+    });
+
+    it('should not change any other values', () => {
+      mount();
+      CredentialsLogic.actions.setCredentialsDetails(credentialsDetails);
+      expect(CredentialsLogic.values).toEqual({
+        ...DEFAULT_VALUES,
+        engines: expect.any(Array),
+        apiUrl: expect.any(String),
+        isCredentialsDetailsComplete: expect.any(Boolean),
+      });
+    });
+  });
+
+  describe('setNameInputBlurred', () => {
+    describe('nameInputBlurred', () => {
+      it('should set this value', () => {
+        mount({
+          nameInputBlurred: false,
+        });
+        CredentialsLogic.actions.setNameInputBlurred(true);
+        expect(CredentialsLogic.values.nameInputBlurred).toEqual(true);
+      });
+    });
+
+    it('should not change any other values', () => {
+      mount();
+      CredentialsLogic.actions.setNameInputBlurred(true);
+      expect(CredentialsLogic.values).toEqual({
+        ...DEFAULT_VALUES,
+        nameInputBlurred: expect.any(Boolean),
+      });
+    });
+  });
+
+  describe('setTokenReadWrite', () => {
+    describe('activeApiToken', () => {
+      // TODO should probably have a more specific behavior
+      it('sets an arbitrary key and value on the activeApiToken', () => {
```

Comment:
`setTokenReadWrite` I think was intended to just change the `read` and `write` properties of a token, but it actually capable of setting *any* property on a token.

Reply:
IMO, we should change this via Typescript in the actions definition and update this test:

```js
interface IReadWriteToken {
  name: 'read' | 'write';
  checked: boolean;
}

export interface ICredentialsLogicActions {
  // ...
  setTokenReadWrite({ name, checked }: IReadWriteToken): IReadWriteToken;
  // ...
}
```

Reply:
a83ca063da569a3dced859645c47d1d3fdea2189